Clean up on the List and List-Method Isles (#2404)
exercism/python: concepts/list-methods/about.md

# About

> A [`list`][list] is a mutable collection of items in _sequence_. Like most collections (_see the built-ins [`tuple`][tuple], [`dict`][dict] and [`set`][set]_), lists can hold reference to any (or multiple) data type(s) - including other lists.  They can be copied in whole or in part via [slice notation][slice notation]. Like any [sequence][sequence type], elements within `lists` are referenced by `0-based index` number.

Reviewer: ```suggestion
A [`list`][list] is a mutable collection of items in _sequence_. Like most collections (_see the built-ins [`tuple`][tuple], [`dict`][dict] and [`set`][set]_), lists can hold references to any (or multiple) data type(s) - including other lists.  They can be copied in whole or in part via [slice notation][slice notation]. Like any [sequence][sequence type], elements within `lists` are referenced by `0-based index` number.
```

Reviewer: Does it make sense to call out when lists are used vs tuples? Or that lists are usually of a uniform type?

Author: 🤔   I did have an earlier iteration of this that used the terms heterogeneous and homogeneous for data...but worried that would lead to confusion, since homogeneity of _data type_ is not a requirement of either container, even though _use_ for `lists` tends to be homogeneous.  Any thoughts or suggestions on how you'd do that?  Here is what I did for `tuples`:  https://github.com/exercism/python/blob/main/concepts/tuples/about.md#tuples-as-related-information

Reviewer: > By convention, lists are used for collections of items when the items are all similar and there can be any number of those items - like scores or items in a shopping basket. In contract, tuples are used when a piece of data has multiple pieces, like an `(x, y)` or a polar coordinate - `(radius, degree)`.

Author: I like that.  Second question:  do we want that in this `about.md` (which is for list-methods) -- or should we be putting it in the `about.md` for lists?

Reviewer: `lists` seems to be the basics without too many methods and `list-methods` seems like the extended method list. As such, I think `list-methods` makes more sense.

> Lists support both [common][common sequence operations] and [mutable][mutable sequence operations] sequence opterations like `min()`/`max()`, `<list>.index()`, `.append()` and `.reverse()`. Items can be iterated over using the `for item in <list>` construct, and `for item in enumerate(<list)` can be used when both and item value and index are needed.

Reviewer: ```suggestion
Lists support both [common][common sequence operations] and [mutable][mutable sequence operations] sequence opterations like `min(<list>)`/`max(<list>)`, `<list>.index()`, `<list>.append()` and `<list>.reverse()`. Items can be iterated over using the `for item in <list>` construct, and `for item in enumerate(<list)` can be used when both the element and index are needed.
```

Python provides many useful [methods][list-methods] for working with lists. Let's take a look at some of them.

> Keep in mind that when you manipulate a list with a list-method, **you alter the list** object that has been passed. If you do not wish to mutate your original `list`, you will need to at least make a `shallow copy` of it via slice or `<list>.copy()`.

Reviewer: You reference "that has been passed" but functions and passing isn't part of this context.
Drop the "that has been passed" or explicitly name "inside a function"

Author: Yup.  Added "into the method" as a qualifier.  Thanks for the catch.  😄

Author: you should be able to see an update.

> Keep in mind that when you manipulate a list with a list-method, **you alter the list** object that has been passed. If you do not wish to mutate your original `list`, you will need to at least make a `shallow copy` of it via slice or `<list>.copy()`.

Author: ```suggestion
Keep in mind that when you manipulate a list with a list-method, **you alter the list** object that has been passed into the method. If you do not wish to mutate your original `list`, you will need to at least make a `shallow copy` of it via slice or `<list>.copy()`.
```

> Keep in mind that when you manipulate a list with a list-method, **you alter the list** object that has been passed. If you do not wish to mutate your original `list`, you will need to at least make a `shallow copy` of it via slice or `<list>.copy()`.

Author: ```suggestion
Keep in mind that when you manipulate a list with a list-method, **you alter the list** object that has been passed into that list-method. If you do not wish to mutate your original `list`, you will need to at minimum make a `shallow copy` via slice or `<list>.copy()`.
```

[3, 2, 1]
```

> List elements can also be sorted  _**in place**_ with the help of `list.sort()`. Internally, Python uses [`Timsort`][timsort] to arrange elements. The default order is _ascending_. Take a look at the Python docs for some [additional tips and techniques for sorting][sorting how to] lists effectively.

Reviewer: ```suggestion
List elements can also be sorted  _**in place**_ using `list.sort()`. Internally, Python uses [`Timsort`][timsort] to arrange elements. The default order is _ascending_. Take a look at the Python docs for some [additional tips and techniques for sorting][sorting how to] lists effectively.
```

Reviewer: Are you using `<list>.sort()` or `list.sort()`? It is the former on L106.

Author: `<list>.sort()` to imply that `<list>` be replaced with the name/list in question.  But if the usage is unclear then we should change it.

Author: Committing this so I can then re-edit it.

["Bruce", "Natasha", "Thor", "Tony"]
```

> If you want the sort to be in _descending_ order, pass a `reverse=True` argument:

Reviewer: ```suggestion
If you want to sort a list in _descending_ order, pass a `reverse=True` argument:
```

["Tony", "Thor", "Natasha", "Bruce"]
```

> For cases where mutating the original list is undesirable, the built-ins [`sorted()`][sorted] and [`reversed()`][reversed] can be used.  `sorted()` will return a sorted copy, and takes the same parameters as `<list>.sort()`. `reversed()` will return a reverse _iterator_ that can be used to build a new list copy.

Reviewer: ```suggestion
For cases where mutating the original list is undesirable, the built-in functions [`sorted()`][sorted] and [`reversed()`][reversed] can be used.  `sorted()` will return a sorted copy, and takes the same parameters as `<list>.sort()`. `reversed()` will return a reverse _iterator_ that can be used to build a new list copy.
```

Reviewer: Is `iterator` going to confuse students if we're assuming they don't yet know what that is?

Author: It might.  But I don't know how else we'd refer to it.  We can remove both that and the ref to `reversed()` -- we could always introduce it later.  You recommendation?  Would someone coming to Python from another programming language be confused by `iterator`??

Reviewer: Maybe,

`reversed()` returns an _iterator_ that yields the list's items in reverse order. We'll talk about _iterators_ more in later exercises.

Author: Nice!  I like that.  😄

> For cases where mutating the original list is undesirable, the built-ins [`sorted()`][sorted] and [`reversed()`][reversed] can be used.  `sorted()` will return a sorted copy, and takes the same parameters as `<list>.sort()`. `reversed()` will return a reverse _iterator_ that can be used to build a new list copy.

Author: ```suggestion
For cases where mutating the original list is undesirable, the built-in functions [`sorted()`][sorted] and [`reversed()`][reversed] can be used.  `sorted(<list>)` will return a sorted _copy_, and takes the same parameters as `<list>.sort()`. `reversed(<list>)` returns an _iterator_ that yields the list's items in reverse order. We'll talk about iterators more in later exercises.
```

## Making Copies

> Remember that variables in Python are labels that point to underlying objects and `lists` are _container_ objects that only hold references to their collected items.

Reviewer: s/collected/contained/ ?

Author: dunno.  They feel interchangeable to me.  Do you have  a preference?

Reviewer: "collected" feels weird to me. Containers don't collect things. A person might collect things and put them into a container. But containers *do* contain objects.

Author: Good point.  I think I used "collected" because a list is a _type_ of _collection_.  But weird is weird.  Let's go with contain.  😄

> Remember that variables in Python are labels that point to underlying objects and `lists` are _container_ objects that only hold references to their collected items.

Author: ```suggestion
Remember that variables in Python are labels that point to underlying objects and `lists` are _container_ objects that only hold references to their contained items.
```

> Assigning a `list` object to a new variable _name_ does not copy the object or any of its refrenced data. Any change made to the items in the `list` using the new variable name will also _impact the original_.

Reviewer: ```suggestion
Assigning a `list` object to a new variable _name_ does not copy the object or any of its referenced data. Any change made to the items in the `list` using the new variable name will also _impact the original_.
```

> `<list>.copy()` will create a new `list` object, but **will not** create new objects for the referenced list _elements_. This type of copy is usually enough to add or remove items from  two `list` objects independantly and avoid problems. But if there is a chance that the elements of the `list` might be accidentially mutated and a full copy of **all** refrences and objects is needed, [`copy.deepcopy()`][deepcopy] in the `copy` module should be used.

Reviewer: ```suggestion
`<list>.copy()` will create a new `list` object, but **will not** create new objects for the referenced list _elements_. This is called a shallow copy. This type of copy is usually enough when you want to add or remove items from one of the `list` objects without modifying the other. But if there is a chance that the elements of the `list` might accidentally be mutated, a full copy of **all** references and objects should be use. [`copy.deepcopy()`][deepcopy] in the `copy` module should be used for this.
```

Author: oof.  Decided to do a little re-arrange.  Self-suggested/edited.

> `<list>.copy()` will create a new `list` object, but **will not** create new objects for the referenced list _elements_. This type of copy is usually enough to add or remove items from  two `list` objects independantly and avoid problems. But if there is a chance that the elements of the `list` might be accidentially mutated and a full copy of **all** refrences and objects is needed, [`copy.deepcopy()`][deepcopy] in the `copy` module should be used.

Author: ```suggestion
`<list>.copy()` will create a new `list` object, but **will not** create new objects for the referenced list _elements_. This is called a `shallow copy`.  A `shallow copy` is usually enough when you want to add or remove items from one of the `list` objects without modifying the other.  But if there is any chance that the _underlying_ elements of a `list` might be accidentally mutated (_thereby mutating all related shallow copies_),  [`copy.deepcopy()`][deepcopy] in the `copy` module should be used to create a complete or "deep" copy of **all** references and objects.
```
